src/crimson/osd/pg.cc

```diff
@@ -141,7 +140,8 @@ bool PG::try_flush_or_schedule_async() {
 void PG::queue_check_readable(epoch_t last_peering_reset, ceph::timespan delay)
 {
   // handle the peering event in the background
-  std::ignore = seastar::sleep(delay).then([last_peering_reset, this] {
+  check_readable_timer.cancel();
```

Comment:
could you 

- add a `future<> PG::stop()` method. and 
- cancel all timers in it
- and do something like 
```c++
seastar::do_for_each(pg_map.get_pgs(), [](auto& pg) {
  return pg.second->stop();
});
```

otherwise, we could have dangling `this` pointer referencing a PG in a unfinished task for posting this event to the pg in question when shutting down OSD.

we could do this in parallel, but i think it's not a bottleneck when stopping an OSD. and it's simpler to use `seastar::do_for_each` than using `seastar::parallel_for_each()`.

Reply:
> but since eventually, since the timestamp in the more updated message will overwrite the old one. probably we'd better to cancel the old events if they are not sent yet.

Yes, this is what I've observed from OSD when I changed pool size. There is another event of `schedule_renew_lease()` with a newer epoch. And I have to cancel the pending one which is outdated, because `seastar::timer` doesn't allow to keep different callbacks in the same timer.
 So I assume this cancel logic is correct.

> otherwise, we could have dangling this pointer referencing a PG in a unfinished task for posting this event to the pg in question when shutting down OSD.

I agree that `future<> PG::stop()` is necessary for a clean shutdown, but I don't think we need it for this patch with `seastar::timer`, as it is RAII alike and will cancel itself during destruction.